Azure OpenAI: 2.0.0-beta.3 updates

ralph-msft opened the pull request:
Also includes the following changes:
- Moved to the latest 2024-07-01-preview for Azure OpenAI
- Updated test recordings

Commit:
Code compiles with external OpenAI library

Commit:
Most tests pass and have been re-recorded against the latest service version.

The tests that currently fail are all related to the SerializedAdditionalRawData not being popualated regression.

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.AI.OpenAI](https://apiview.dev/Assemblies/Review/c2e06b19612a4b8690a3230e8a1cd596?revisionId=3773fccdc2684cd590fb9cf8e01ab492)

trrwilson commented on sdk/openai/Azure.AI.OpenAI/src/Azure.AI.OpenAI.csproj:
```diff
@@ -11,13 +11,20 @@
   </PropertyGroup>
 
-  <ItemGroup>
+  <ItemGroup Condition="'$(UseExternalLib)' == 'true'">
```
It'd be good to leave a brief, explanatory `<!-- comment -->` here (also in the Directory.build.props, which could then accompany the commented-out setting to true with the information about when to uncommented it for local validation)

ralph-msft replied:
I've refactored this a little and added more comments in another place for more context on why this was added. I will push out another update soon.

trrwilson commented on sdk/openai/Azure.AI.OpenAI/src/Custom/Assistants/AzureRunStepDetailsUpdate.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Azure.AI.OpenAI;
+
+public static class RunStepDetailsUpdateExtensions
```
ah, I missed this -- we need to remove (or at least [EditorBrowsable(Never)]) these Assistants files since the matching feature was briefly postponed.

ralph-msft replied:
I'll remove all these types.

trrwilson commented on sdk/openai/Azure.AI.OpenAI/src/Custom/Assistants/Internal/GeneratorStubs.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Azure.AI.OpenAI.Assistants;
+
+[CodeGenModel("BingSearchToolDefinitionBrowser")] internal partial class InternalBingSearchToolDefinitionBrowser { }
```
Per the above, all these types don't exist anymore (for now) -- the spec and generated code got the update, but I missed the custom code for the extension methods.

ralph-msft replied:
I removed this file outrigt.

trrwilson commented on sdk/openai/external/lib/netstandard2.0/OpenAI.dll:
It's likely just temporary anyhow, but we shouldn't permanently put the static binary in the repository -- if we wanted to keep it as a resource (I suspect we wouldn't and instead would want to set up the alpha feed), it'd fit better into assets.

ralph-msft replied:
This was a mistake on my part. These should not be checked in and I've deleted these files and added a .gitignore for this  as part of my next update.

Commit:
Address PR feedback
- Cleaned up the asssembly instead of nuget reference tweaks
- Removed unnecessary types

Commit:
Fix bug in adapter. Clean up how to add custom OpenAI assembly, and support for custom Azure.Core assembly.

Commit:
Remove unneeded file

Commit:
Update test recordings

chschrae commented on sdk/openai/Azure.AI.OpenAI/src/Custom/Assistants/AzureAssistantClient.Protocol.cs:
```diff
@@ -107,22 +107,21 @@ public override ClientResult CreateMessage(string threadId, BinaryContent conten
 
-        using PipelineMessage message = CreateGetMessagesRequest(threadId, limit, order, after, before, options);
-        return ClientResult.FromResponse(await Pipeline.ProcessMessageAsync(message, options).ConfigureAwait(false));
+        AzureMessagesPageEnumerator enumerator = new(Pipeline, _endpoint, threadId, limit, order, after, before, _apiVersion, options);
```
should these be using `using`?

ralph-msft replied:
I believe so, since we don't want to hold onto any networking streams since we process them here

joseharriaga committed:
Remove constructors that do not explicitly take a credential parameter (#201)

- Removed client constructors that do not explicitly take an API key parameter or an endpoint via an `OpenAIClientOptions` parameter, making it clearer how to appropriately instantiate a client.
- Removed the endpoint parameter from all client constructors, making it clearer that an alternative endpoint must be specified via the `OpenAIClientOptions` parameter.
- Removed `OpenAIClient`'s `Endpoint` `protected` property.
- Made `OpenAIClient`'s constructor that takes a `ClientPipeline` parameter `protected internal` instead of just `protected`.

Commit:
Update .csprof files to make it easier to change where our Auzre.Core, and OpenAI depencies come from.

Commit:
Update based on latest OpenAI changes

Commit:
Update test recordings since JSON serialization order has changed

Commit:
Update to latest OpenAI Nuget package version.

trrwilson committed:
changelog, readme, api updates for release

trrwilson committed:
minor readme re-merge

Commit:
Merge remote-tracking branch 'origin/main' into ralphe/update_oai_version

Commit:
Merge from main. ChateTest.RateLimitedRetryWorks is failing. All other tests pass.

Commit:
Merge against latest from main

Commit:
Nothing to see here, move along
